File: 04-linked-lists/07_cycle_start.c
#include "../test_runner.h"

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

ListNode* make_node(int val) {
    ListNode* n = malloc(sizeof(ListNode));
    n->val = val; n->next = NULL;
    return n;
}

ListNode* make_list(int* arr, int n) {
    if (n == 0) return NULL;
    ListNode* head = make_node(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; i++) {
        curr->next = make_node(arr[i]);
        curr = curr->next;
    }
    return head;
}

void to_array(ListNode* head, int* out, int* len) {
    *len = 0;
    while (head) { out[(*len)++] = head->val; head = head->next; }
}


ListNode* detect_cycle(ListNode* head) {
    return NULL;
}

int main() {
    ListNode* n1=make_node(3); ListNode* n2=make_node(2);
    ListNode* n3=make_node(0); ListNode* n4=make_node(-4);
    n1->next=n2; n2->next=n3; n3->next=n4; n4->next=n2;
    TEST_INT("cycle at 2", detect_cycle(n1)->val, 2);
    TEST_BOOL("no cycle", detect_cycle(make_node(1)) == NULL, true);
    TEST_SUMMARY();
}

File: 04-linked-lists/07_cycle_start.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests



# Linked List Cycle II (LC #142) -- Medium
# Input:  head: ListNode (may have cycle)
# Output: ListNode (node where cycle starts, or None)
# If linked list has a cycle, return the NODE where cycle begins.
# Return None if no cycle.
#
# Example:
#   3 -> 2 -> 0 -> -4 -> (back to 2)  -> return node with val=2
#   1 -> 2 -> (back to 1)              -> return node with val=1
#   1 -> None                          -> return None
#
#   Key: Floyd's algorithm, phase 2.
#        1. Detect cycle with slow/fast.
#        2. Move one pointer to head, keep other at meeting point.
#        3. Both advance 1 step. They meet at cycle start.

class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


def make_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def to_array(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr


def detect_cycle(head):
    """Return the node where cycle begins, or None"""
    pass


def solve(arr, pos):
    head = make_list(arr)
    if pos == -1:
        node = detect_cycle(head)
        return -1 if node is None else node.val
    nodes = []
    curr = head
    while curr:
        nodes.append(curr)
        curr = curr.next
    nodes[-1].next = nodes[pos]
    node = detect_cycle(head)
    return -1 if node is None else node.val


run_tests(solve, [
    (([3,2,0,-4], 1),   2),     # cycle starts at node val=2
    (([1,2], 0),        1),     # cycle starts at head
    (([1], -1),        -1),     # no cycle
    (([1,2,3,4,5], 2),  3),     # cycle starts at node val=3
    (([1], 0),          1),     # self-loop
])
